src/main/cpp/src/hyper_log_log_plus_plus_host_udf.cu

```diff
@@ -0,0 +1,173 @@
+  {
+    // groupby
+    auto const& group_values = get_grouped_values();
+    if (group_values.size() == 0) { return get_empty_output(stream, mr); }
+    int num_groups          = get_num_groups();
+    auto const group_labels = get_group_labels();
+    if (is_merge) {
+      // group by intermidate result, group_values are struct of long columns
+      return spark_rapids_jni::group_merge_hyper_log_log_plus_plus(
+        group_values, num_groups, group_labels, precision, stream, mr);
+    } else {
+      return spark_rapids_jni::group_hyper_log_log_plus_plus(
+        group_values, num_groups, group_labels, precision, stream, mr);
+    }
+  }
+
+  /**
+   * @brief Create an empty column when the input is empty.
+   */
+  [[nodiscard]] std::unique_ptr<cudf::column> get_empty_output(
+    rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr) const override
+  {
+    int num_registers       = 1 << precision;
+    int num_long_cols       = num_registers / REGISTERS_PER_LONG + 1;
+    auto const results_iter = cudf::detail::make_counting_transform_iterator(
+      0, [&](int i) { return cudf::make_empty_column(cudf::data_type{cudf::type_id::INT64}); });
+    auto children =
+      std::vector<std::unique_ptr<cudf::column>>(results_iter, results_iter + num_long_cols);
+    return cudf::make_structs_column(0,
+                                     std::move(children),
+                                     0,                     // null count
+                                     rmm::device_buffer{},  // null mask
+                                     stream,
+                                     mr);
+  }
+
+  [[nodiscard]] bool is_equal(cudf::host_udf_base const& other) const override
+  {
+    auto o = dynamic_cast<hllpp_agg_udf const*>(&other);
+    return o != nullptr && o->precision == this->precision && o->is_merge == this->is_merge;
+  }
+
+  [[nodiscard]] std::size_t do_hash() const override
+  {
+    return 31 * (31 * std::hash<std::string>{}({"hllpp_agg_udf"}) + precision) + is_merge;
+  }
+
+  [[nodiscard]] std::unique_ptr<cudf::host_udf_base> clone() const override
+  {
+    return std::make_unique<hllpp_agg_udf>(precision, is_merge);
+  }
+
+  int precision;
+  bool is_merge;
+};
+
+struct hllpp_reduct_udf : cudf::reduce_host_udf {
+  hllpp_reduct_udf(int precision_, bool is_merge_) : precision(precision_), is_merge(is_merge_) {}
+
+  /**
+   * Perform the main reduce computation for HLLPP UDF
+   */
+  std::unique_ptr<cudf::scalar> operator()(
+    cudf::column_view const& input,
+    cudf::data_type,                                           /** output_dtype is useless */
+    std::optional<std::reference_wrapper<cudf::scalar const>>, /** init is useless */
+    rmm::cuda_stream_view stream,
+    rmm::device_async_resource_ref mr) const override
+  {
+    CUDF_EXPECTS(input.size() > 0,
+                 "Hyper Log Log Plus Plus reduction requires input is not empty!");
```

Comment:
Why? Can we return an empty output instead?

Reply:
done,  add a function `get_empty_scalar`